### PR TITLE
MdeModulePkg: Add InsecureDeviceState PCD

### PR DESCRIPTION
**Pull request by VivianNK**
## Description

Add PCD to provide a FixedAtBuild value to define what an insecure Device state is.

For example, the default value defines an insecure device state as a DEVICE_STATE with DEVICE_STATE_SOURCE_DEBUG_ENABLED, or DEVICE_STATE_UNIT_TEST_MODE.

Add a function to the DeviceStateLib, GetInsecureDeviceStateSettings, that retrieves the PCD. Expected use cases include comparing this PCD against current device state to check for an insecure device state, then measuring into the TPM.

This PR has a test that's dependent on a mock for PcdLib being added in MdePkg https://github.com/microsoft/mu_basecore/pull/1178

- [ ] Impacts functionality?
- [x] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [x] Includes documentation?

## How This Was Tested

DeviceStateLibGoogleTest.inf

## Integration Instructions

Update the FixedAtBuild PCD InsecureDeviceState to include any device states that the platform considers insecure. Call this API to take action if an insecure device state has been entered.

Expected use cases include comparing this bitmask against the current device state to check
for an insecure device state, then measuring into the TPM. 

Make any modifications needed to accommodate all OEM-defined device states. 

To add the measurement of this PCD in Tcg2Pei, update your mu_tiano_plus submodule to the release/202405 branch or release tag once this change is merged (https://github.com/microsoft/mu_tiano_plus/pull/342).

### CODE REVIEW

File: MdeModulePkg/Include/Library/DeviceStateLib.h
Comment (os-d) [inline MdeModulePkg/Include/Library/DeviceStateLib.h]: I think this can just return a `DEVICE_STATE` since it is being used to compare against a `DEVICE_STATE`.

File: MdeModulePkg/Library/DeviceStateLib/Readme.md
Comment (os-d) [inline MdeModulePkg/Library/DeviceStateLib/Readme.md]: Can you add some more details here on the expected use case for this (i.e. measuring in to the TPM)

File: MdeModulePkg/MdeModulePkg.dec
Comment (os-d) [inline MdeModulePkg/MdeModulePkg.dec]: This is the wrong format. It should be `Name|Value|Type|UniqueIdentifier`

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (os-d) [inline MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: Let's use `FixedPcdGet32`

File: MdeModulePkg/Library/DeviceStateLib/Readme.md
Comment (os-d) [reply MdeModulePkg/Library/DeviceStateLib/Readme.md]: Can you similarly add the intended use case for this to the PR description?

File: MdeModulePkg/Include/Library/DeviceStateLib.h
Comment (cfernald) [inline MdeModulePkg/Include/Library/DeviceStateLib.h]: Maybe GetInsecureDeviceStateSettings or something that conveys what this returns a bit better.

File: MdeModulePkg/Library/DeviceStateLib/Readme.md
Comment (VivianNK) [reply MdeModulePkg/Library/DeviceStateLib/Readme.md]: Updated ReadMe

File: MdeModulePkg/Include/Library/DeviceStateLib.h
Comment (os-d) [inline MdeModulePkg/Include/Library/DeviceStateLib.h]: I missed this before. Can we also not call out the PCD here? This should just describe the interface, not implementation details, so that if the library is overridden, it can change the implementation as it needs. And then this comment should be exactly the same as in the c file.

So something like

```c
/**
  Get the platform defined set of insecure device states. This can be used to measure insecure device
  states into the TPM or perform other required platform actions when the device enters an insecure state.

  @retval  The bitmask of insecure device states as defined by the platform.
**/
```

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (os-d) [inline MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: nit: You can just return this directly.

File: MdeModulePkg/Library/DeviceStateLib/Readme.md
Comment (os-d) [inline MdeModulePkg/Library/DeviceStateLib/Readme.md]:
```suggestion
gMdeModulePkgTokenSpaceGuid.InsecureDeviceState. This PCD, by default, ORs DEVICE_STATE_SOURCE_DEBUG_ENABLED with DEVICE_STATE_UNIT_TEST_MODE, but can be customized to include other states.
```

File: MdeModulePkg/Library/DeviceStateLib/Readme.md
Comment (os-d) [inline MdeModulePkg/Library/DeviceStateLib/Readme.md]:
```suggestion
Expected use cases include comparing this bitmask against the current device state to check for an insecure device state, then measuring into the TPM. A platform can update this PCD with any other states it considers to be insecure and can call this API to take action if an insecure device state has been entered.
```

File: MdeModulePkg/MdeModulePkg.dec
Comment (os-d) [inline MdeModulePkg/MdeModulePkg.dec]:
```suggestion
  # Default value is DEVICE_STATE_SOURCE_DEBUG_ENABLED | DEVICE_STATE_UNIT_TEST_MODE as defined in Include/Library/DeviceStateLib.h

```

**Review by os-d — APPROVED** [review]

File: MdeModulePkg/Include/Library/DeviceStateLib.h
Comment (os-d) [inline MdeModulePkg/Include/Library/DeviceStateLib.h:38]: Is this from the TCG spec?

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (os-d) [inline MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: Where possible, we should avoid UINTN. It should only be used in a case where we actually need a different size on different architectures (like for computing integers as pointers.

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (os-d) [inline MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: nit: none of the parentheses are needed

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (os-d) [inline MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: This API doesn't seem to add value. We can drop it and just directly do the bitwise & here.

File: MdeModulePkg/Include/Library/DeviceStateLib.h
Comment (os-d) [inline MdeModulePkg/Include/Library/DeviceStateLib.h:49]: I assume in Tcg2Pei you will also measure the specific TCG spec defined debugger string?

**Review by os-d — APPROVED** [review]

File: MdeModulePkg/Include/Library/DeviceStateLib.h
Comment (VivianNK) [reply MdeModulePkg/Include/Library/DeviceStateLib.h:49]: https://github.com/microsoft/mu_tiano_plus/pull/342/files#diff-dd252bb953ec11695bb497d99150fe3c48e069945f08a98630c4a2e47baef3b3R660
yes

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (VivianNK) [reply MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: I added it for readability, specifically for consumption in Tcg2Pei. (https://github.com/microsoft/mu_tiano_plus/pull/342/files#diff-dd252bb953ec11695bb497d99150fe3c48e069945f08a98630c4a2e47baef3b3R1169)
But I can remove it if it's not achieving that goal/it's not much better than bitwise operation. I can add to the function descriptions that use it instead

File: MdeModulePkg/Include/Library/DeviceStateLib.h
Comment (VivianNK) [reply MdeModulePkg/Include/Library/DeviceStateLib.h:38]: Discussed offline, there's no TCG spec defined max

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (VivianNK) [reply MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: dropped

File: MdeModulePkg/Include/Library/DeviceStateLib.h
Comment (makubacki) [inline MdeModulePkg/Include/Library/DeviceStateLib.h:82]: Since this function is at the interface level, it should state the string format returned as a contract to the caller. For example, the string convention ("macro case"), the separation character (e.g. space), etc.

Although someone could postulate that the macros above which are "macro case" and contain space separators are concatenated together based on looking at the entirety of this file that's different from an explicitly stated interface requirement.

File: MdeModulePkg/Include/Library/DeviceStateLib.h
Comment (makubacki) [inline MdeModulePkg/Include/Library/DeviceStateLib.h]: The interface should explicitly describe the behavior taken when `MaxSize` insufficient to hold the string.

File: MdeModulePkg/Include/Library/DeviceStateLib.h
Comment (makubacki) [inline MdeModulePkg/Include/Library/DeviceStateLib.h]: You can drop the excess asterisks. For example:

https://github.com/microsoft/mu_basecore/blob/a45519aa3d4e033caed186eff927557381282bf4/MdePkg/Include/Library/BaseMemoryLib.h#L16-L32

File: MdeModulePkg/Include/Library/DeviceStateLib.h
Comment (makubacki) [inline MdeModulePkg/Include/Library/DeviceStateLib.h]: The function behavior when an invalid parameter is given should be defined.

For example:

- What is the expected behavior if `Buffer` is `NULL`?
- What is the expected behavior if `MaxSize` is `0`?

Comment (makubacki) [inline MdeModulePkg/Include/Library/DeviceStateLib.h]: I suggest using `@return` here versus `@retval`.

From EDK II C Coding Standards Specification:
image

An example that interestingly uses both, but the character count is `@return`:

https://github.com/microsoft/mu_basecore/blob/a45519aa3d4e033caed186eff927557381282bf4/MdePkg/Include/Library/BaseLib.h#L1027-L1037

File: MdeModulePkg/Include/Library/DeviceStateLib.h
Comment (makubacki) [inline MdeModulePkg/Include/Library/DeviceStateLib.h]: Another case where I think `@return` makes sense.

File: MdeModulePkg/Include/Library/DeviceStateLib.h
Comment (makubacki) [inline MdeModulePkg/Include/Library/DeviceStateLib.h]: No need to precede with "Function to..".

```suggestion
 Get the current device state.
```

File: MdeModulePkg/Include/Library/DeviceStateLib.h
Comment (makubacki) [inline MdeModulePkg/Include/Library/DeviceStateLib.h]: Per EDK II C Coding Standards Specification.

image

```suggestion
GetDeviceState (
  VOID
  );
```

File: MdeModulePkg/Include/Library/DeviceStateLib.h
Comment (makubacki) [inline MdeModulePkg/Include/Library/DeviceStateLib.h]:
```suggestion
/**
  Add an additional device state.
  
  @param[in]  AdditionalState    The additional device to state to add.
  
  @retval  EFI_SUCCESS           The device state was added successfully.
**/
```

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (makubacki) [inline MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: I suggest making `6` a macro defined value for clarity. For example, the earlier code defined `7` strings:

```c
#define DEVICE_STATE_SECUREBOOT_OFF_STRING             "SECURE_BOOT_OFF "
#define DEVICE_STATE_MANUFACTURING_MODE_STRING         "MANUFACTURING_MODE "
#define DEVICE_STATE_DEVELOPMENT_BUILD_ENABLED_STRING  "DEVELOPMENT_BUILD_ENABLED "
#define DEVICE_STATE_SOURCE_DEBUG_ENABLED_STRING       "SOURCE_DEBUG_ENABLED "
#define DEVICE_STATE_UNDEFINED_STRING                  "UNDEFINED "
#define DEVICE_STATE_UNIT_TEST_MODE_STRING             "UNIT_TEST_MODE "
#define DEVICE_STATE_PLATFORM_MODE_STRING              "OEM_DEFINED "
```

So, if the macro name indicated why these 6 were chosen and `DEVICE_STATE_PLATFORM_MODE_STRING` was excluded that would make it clear it was intentional and how to determine whether future strings should be added here or not.

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (makubacki) [inline MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: Need to handle the case where `Buffer` is `NULL` to prevent null pointer dereference.

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (makubacki) [inline MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: Typically, loop indices are named `Index`. I personally prefer that to `i` for consistency with the rest of the codebase.

EDK II C Coding Standards Specification example:

image

Comment (makubacki) [reply MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: I'm not aware of an actual requirement here though. If you prefer `i`, please keep it.

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (makubacki) [inline MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: Please use `ARRAY_SIZE`.

```suggestion
  UINT32  NumberOfDeviceStates = ARRAY_SIZE (DEVICE_STATE_STRINGS);
```

Comment (makubacki) [inline MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: If there is not a compelling reason for `NumberOfDeviceStates` to be a `UINT32`, it should be a `UINTN`.

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (makubacki) [inline MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: We typically don't declare variables later in functions. For consistency, I suggest placing the variable declaration at the beginning of the function.

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (makubacki) [inline MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: The loop index should also be of type `UINTN` unless there is a strong need otherwise since it sized to the pointer size of the platform and the index is used as an offset from the base address of the array.

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (makubacki) [inline MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: A check to make sure the shift operation of `1 << i` is within range of `BitmaskFromIndex` should occur before this (i.e. `DEVICE_STATE_MAX` specifies the maximum as `31`).

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (makubacki) [inline MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: If I understand correctly, I suggest calling this variable `NumberOfDeviceStateStrings`.

```suggestion
  UINT32  NumberOfDeviceStateStrings = sizeof (DEVICE_STATE_STRINGS) / sizeof (DEVICE_STATE_STRINGS[0]);
```

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (makubacki) [inline MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: Was there a reason not to call `AsciiStrnSizeS()`? It includes the null character taking into account platform character width.

https://github.com/microsoft/mu_basecore/blob/a45519aa3d4e033caed186eff927557381282bf4/MdePkg/Library/BaseLib/SafeString.c#L1762-L1770

```suggestion
  return AsciiStrnSizeS (Buffer, MaxSize);
```

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (makubacki) [inline MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: Is a function definition missing between these comments?

```c
/**
 * Check if a specific bitmask is set in the device state
 *
 * @param[in] CurrentDeviceState - the current device state
 * @param[in] DeviceStateBitmask - the bitmask to check
 * @retval TRUE if all the bits in the bitmask are set, FALSE otherwise
 */

/**
 * Function to Get current device state
 *
 * @retval the current DEVICE_STATE
 */
```

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (makubacki) [inline MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c:2]: Since this talks about `PcdDeviceStateBitmask`, you might also want to describe how this library instance uses `PcdInsecureDeviceState` (i.e. fixed PCD).

File: MdeModulePkg/Include/Library/DeviceStateLib.h
Comment (makubacki) [inline MdeModulePkg/Include/Library/DeviceStateLib.h]: Not a big deal, but personal preference is to directly associate the values with the reason they were chosen.

```suggestion
#define DEVICE_STATE_PLATFORM_MODE_INDEX_START  DEVICE_STATE_PLATFORM_MODE_0
#define DEVICE_STATE_PLATFORM_MODE_INDEX_END    DEVICE_STATE_PLATFORM_MODE_7
```

File: MdeModulePkg/Include/Library/DeviceStateLib.h
Comment (makubacki) [inline MdeModulePkg/Include/Library/DeviceStateLib.h]:
```suggestion
// These strings are used to represent device state for logging and measurements.
```

File: MdeModulePkg/Include/Library/DeviceStateLib.h
Comment (makubacki) [inline MdeModulePkg/Include/Library/DeviceStateLib.h:1]: The integration instructions for the PR should have details for platforms to understand how to integrate and use this change (it is currently `N/A`).

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (os-d) [reply MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: Ha, for this and the loop index below Vivian had them as UINTN and I said if there isn’t a compelling reason to have them UINTN, they should be a fixed type. I think we see a lot of weirdness around UINTN, generally not for a good reason. But, I defer to your judgement

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (os-d) [reply MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: This I agree though, that since Index (or i) is being used in pointer arithmetic, it makes sense to have as UINTN and then since it is being compared to NumberOfDeviceStates that might as well be UINTN, too. Sorry for the noise, Vivian.

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (makubacki) [reply MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: Yeah, `UINTN` is generally best for counts and indices because those tend to get used in pointer arithmetic and represent values that scale with pointer width. For example, `UINTN` on a 64-bit machine can accurately represent the number of bytes for a valid offset where that offset can be larger than `MAX_UINT32`.

There are of course cases where smaller integers make sense. Especially in firmware, we have fixed size fields in data structures irrespective of the current architecture. In that case, the maximum allowable value is constrained by the space allocated for the data rather than the architecture.

In both cases, it can simplify casting and result in overall better integer handling to apply the appropriate type for the use case. In this case, I see count and indexing into an array being used which is classically better handled with `UINTN`. And you usually need the count and index type to match for their comparison in the loop to be safe.

File: MdeModulePkg/Include/Library/DeviceStateLib.h
Comment (VivianNK) [reply MdeModulePkg/Include/Library/DeviceStateLib.h]: Changed the function prototype to return EFI_STATUS. AddDeviceState uses RETURN_STATUS but is an EFIAPI - should that be changed as well?

Also, what do you think of the IN/OUT UINTN* for StringSize? Presumably the caller can just call AsciiStrnSizeS themselves?

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (VivianNK) [reply MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: Replaced with macro

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (VivianNK) [reply MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: Added a check but shouldn't this be guaranteed from the NumberOfDeviceStates?

File: MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c
Comment (VivianNK) [reply MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c]: Removed the function earlier and this was missed. Cleaned up now.

File: MdeModulePkg/Include/Library/DeviceStateLib.h
Comment (VivianNK) [reply MdeModulePkg/Include/Library/DeviceStateLib.h]: This would change the values 
1 << 20 instead of 20. 
I did make a change to consume the macros for the start and end mode macros... not sure if that helps. Open to suggestions.

**Review by spbrogan — CHANGES_REQUESTED** [review]
I have some questions I would like to discuss before merging.

**Comment by VivianNK** [comment]
Offline discussion determined it wasn't useful to measure other device states through this library.

**Pull request closed without merging**
